modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxManager.java

```diff
@@ -2922,25 +3017,150 @@ public void longTransactionTimeDumpSamplesPerSecondLimit(int limit) {
      */
     public void longOperationsDumpTimeoutDistributed(long longOpsDumpTimeout) {
         broadcastToNodesSupportingFeature(
+            cctx.kernalContext(),
             new LongOperationsDumpSettingsClosure(longOpsDumpTimeout),
+            false,
             DISTRIBUTED_CHANGE_LONG_OPERATIONS_DUMP_TIMEOUT
         );
     }
 
     /**
-     * Broadcasts given job to nodes that support ignite feature.
+     * Returns tx keys collisions dump interval, for additional info check
+     * {@link IgniteSystemProperties#IGNITE_DUMP_TX_COLLISIONS_INTERVAL} description.
+     *
+     * @return Collisions dump interval.
+     */
+    public int collisionsDumpInterval() {
+        return collisionsDumpInterval;
+    }
+
+    /**
+     * Changes tx key collisions dump interval.
+     * For additional info check {@link IgniteSystemProperties#IGNITE_DUMP_TX_COLLISIONS_INTERVAL} description.
+     *
+     * @param collisionsDumpInterval New collisions dump interval or -1 for disabling.
+     */
+    public void collisionsDumpIntervalDistributed(int collisionsDumpInterval) {
+        broadcastToNodesSupportingFeature(
+            cctx.kernalContext(),
+            new TxCollisionsDumpSettingsClosure(collisionsDumpInterval),
+            true,
+            DISTRIBUTED_TX_COLLISIONS_DUMP
+        );
+    }
+
+    /**
+     * Collect queue size per key collisions info.
+     *
+     * @param key Key..
+     * @param queueSize Collisions queue size
+     **/
+    public void pushCollidingKeysWithQueueSize(GridCacheMapEntry key, int queueSize) {
+        keyCollisionsInfo.put(key, queueSize);
+    }
+
+    /** Wrapper for inner collect logic. */
+    private void collectTxCollisionsInfo() {
+        keyCollisionsInfo.collectInfo();
+    }
+
+    /**
+     * Check local and remote candidates queue size.
      *
-     * @param job Ignite job.
-     * @param feature Ignite feature.
+     * @param entry CacheEntry.
      */
-    private void broadcastToNodesSupportingFeature(IgniteRunnable job, IgniteFeatures feature) {
-        ClusterGroup grp = cctx.kernalContext().grid()
-            .cluster()
-            .forPredicate(node -> IgniteFeatures.nodeSupports(cctx.kernalContext(), node, feature));
+    public void detectPossibleCollidingKeys(GridDistributedCacheEntry entry) {
+        int qSize = entry.remoteMvccSnapshot().size();
+
+        try {
```

Comment:
We might introduce an optimization here that will check qSize and if it's `>= COLLISIONS_QUEUE_THRESHOLD` won't even calculate `localCandidates().size()`.